clippy_lints/src/invalid_ref.rs

```diff
@@ -0,0 +1,59 @@
+use rustc::lint::*;
+use rustc::ty;
+use rustc::hir::*;
+use utils::{match_def_path, paths, span_help_and_lint, opt_def_id};
+
+/// **What it does:** Checks for creation of references to zeroed or uninitialized memory.
+///
+/// **Why is this bad?** Creation of null references is undefined behavior.
+///
+/// **Known problems:** None. 
+///
+/// **Example:**
+/// ```rust
+/// let bad_ref: &usize = std::mem::zeroed();
+/// ```
+
+declare_lint! {
+    pub INVALID_REF,
+    Warn,
+    "creation of invalid reference"
+}
+
+const ZERO_REF_SUMMARY: &str = "reference to zeroed memory";
+const UNINIT_REF_SUMMARY: &str = "reference to uninitialized memory";
+
+pub struct InvalidRef; 
+
+impl LintPass for InvalidRef {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INVALID_REF)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidRef {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if_let_chain!{[
+            let ty::TyRef(..) = cx.tables.expr_ty(expr).sty,
+            let ExprCall(ref path, ref args) = expr.node,
```

Comment:
can we have the exprcall/exprpath/argument length checks first? I know it's not the logical way of looking at it, but these are less expensive so we won't be doing an expr_ty (which is a cache miss) on each expr.